### DISCUSSION TITLE
Tutorials on retrain a MobileNet v2 object detection model for Edge TPU

### DISCUSSION THREAD

**Discussion by igor-rud-kr** [AxisCommunications/acap-computer-vision-sdk-examples#46] [Q&A] [answered]
There is a [Retrain SSD MobileNet V1 detector for the Edge TPU (TF1)](https://colab.research.google.com/github/google-coral/tutorials/blob/master/retrain_detection_qat_tf1.ipynb) tutorial, also there is [Retrain MobileNet V2 classifier for the Edge TPU (TF2)](https://colab.research.google.com/github/google-coral/tutorials/blob/master/retrain_classification_ptq_tf2.ipynb#scrollTo=hRTa3Ee15WsJ) tutorial

Are there any tutorials for MobileNet V2 detector for the Edge TPU retraining?

**Comment by Corallo** [comment, marked as answer]
Hello Igro.

It depends if you are using an artpec7 or artpec8 camera.
If you have a quantized model in the .tflite format, with an artpec7 you can just use the [edgetpu-compiler](https://coral.ai/docs/edgetpu/compiler/#co-compiling-multiple-models) to convert the model and use it on the TPU.
For artpec8 this step is not needed, you can directly load the .tflite model.

For more details about how to retrain and quantize a model, you can also follow these two guides for [artpec7](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/master/tensorflow-to-larod) and [artpec8](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/master/tensorflow-to-larod-artpec8)

**Comment by igor-rud-kr** [comment]
Hi Angelo,

It took me some time to build  tensorflow_to_larod_a8_app_1_0_0_aarch64.eap basing on [ensorflow-to-larod-artpec8](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/tensorflow-to-larod-artpec8) tutorial. I was able to load it to the cameras and judging from /var/volatile/log/info.log log file everything seems to be working ok.

However, now, when I build the application, I don't think this is what I was asking about. I wanted to check if there is a way to use a different model in acap-dl-models module in [Object Detector Python](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/tree/main/object-detector-python) example.

acap-dl-models does the following:

ADD https://github.com/google-coral/edgetpu/raw/master/test_data/ssd_mobilenet_v2_coco_quant_postprocess_edgetpu.tflite models/
ADD https://github.com/google-coral/edgetpu/raw/master/test_data/ssd_mobilenet_v2_coco_quant_postprocess.tflite models/
ADD https://github.com/google-coral/edgetpu/raw/master/test_data/coco_labels.txt models/
ADD https://github.com/google-coral/edgetpu/raw/master/test_data/COPYRIGHT models/

My question is - can I replace these links with my own tflite modles and labels, and if yes, is there a tutorial on how to re-train models to be compatible with acap-dl-models?

**Reply by igor-rud-kr** [reply]
One of the output in [tensorflow-to-larod-artpec8](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/tensorflow-to-larod-artpec8) is a tflite model in app folder - /env/app/coverted_model.tflite. Can I use it in acap-dl-models instead of the two tflite models from google-coral?

**Comment by Corallo** [comment]
Hi Igor

Yes, if you have your own .tflite model, you can just replace those lines in the dockerfile with the path of your model.
From
`ADD https://github.com/google-coral/edgetpu/raw/master/test_data/ssd_mobilenet_v2_coco_quant_postprocess.tflite models/`
to 
`ADD /path/to/my/model.tflite models/`

**Reply by igor-rud-kr** [reply]
Hi Angelo,

Can you please point me in the right direction on how to trace the issue (please see below)

I used the following tutorial:
https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_on_mobile_tensorflowlite.md using ssd_mobilenet_v2_quantized_300x300_coco_2019_01_03.tar.tar (TensorFlow 1). Trying to run it on AXIS P3265-LV.

After I updated object-detector-python example with re-trained model, I received the following error message:


[+] Running 5/5
 ⠿ Network object-detector-python_default                     Created                                                                                                                   0.4s
 ⠿ Volume "object-detector-python_acap_dl-models"             Created                                                                                                                   0.0s
 ⠿ Container object-detector-python-acap_dl-models-1          Created                                                                                                                   1.3s
 ⠿ Container object-detector-python-inference-server-1        Created                                                                                                                   1.3s
 ⠿ Container object-detector-python-object-detector-python-1  Created                                                                                                                   1.3s
Attaching to object-detector-python-acap_dl-models-1, object-detector-python-inference-server-1, object-detector-python-object-detector-python-1
object-detector-python-acap_dl-models-1          | detect3.tflite
object-detector-python-acap_dl-models-1          | labels3.txt
object-detector-python-acap_dl-models-1 exited with code 0
object-detector-python-inference-server-1        | Server listening on 0.0.0.0:8501
object-detector-python-object-detector-python-1  | object-detector-python connect to: inference-server:8501
object-detector-python-object-detector-python-1  | <_InactiveRpcError of RPC that terminated with:
object-detector-python-object-detector-python-1  |      status = StatusCode.DEADLINE_EXCEEDED
object-detector-python-object-detector-python-1  |      details = "Deadline Exceeded"
object-detector-python-object-detector-python-1  |      debug_error_string = "{"created":"@1659657745.590786583","description":"Error received from peer ipv4:172.20.0.4:8501","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"Deadline Exceeded","grpc_status":4}"
object-detector-python-object-detector-python-1  | >
object-detector-python-object-detector-python-1  | Dropping old frames: 3864951
object-detector-python-object-detector-python-1  | Dropping old frames: 3365049
object-detector-python-object-detector-python-1  | Dropping old frames: 2864878
object-detector-python-object-detector-python-1  | <_InactiveRpcError of RPC that terminated with:
object-detector-python-object-detector-python-1  |      status = StatusCode.DEADLINE_EXCEEDED
object-detector-python-object-detector-python-1  |      details = "Deadline Exceeded"
object-detector-python-object-detector-python-1  |      debug_error_string = "{"created":"@1659657750.047190129","description":"Error received from peer ipv4:172.20.0.4:8501","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"Deadline Exceeded","grpc_status":4}"
object-detector-python-object-detector-python-1  | >
object-detector-python-object-detector-python-1  | Dropping old frames: 3797050
object-detector-python-object-detector-python-1  | Dropping old frames: 3297179
object-detector-python-object-detector-python-1  | Dropping old frames: 2797241
object-detector-python-object-detector-python-1  | <_InactiveRpcError of RPC that terminated with:
object-detector-python-object-detector-python-1  |      status = StatusCode.DEADLINE_EXCEEDED
object-detector-python-object-detector-python-1  |      details = "Deadline Exceeded"
object-detector-python-object-detector-python-1  |      debug_error_string = "{"created":"@1659657754.500877746","description":"Error received from peer ipv4:172.20.0.4:8501","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"Deadline Exceeded","grpc_status":4}"
object-detector-python-object-detector-python-1  | >
object-detector-python-object-detector-python-1  | Dropping old frames: 4276186
object-detector-python-object-detector-python-1  | Dropping old frames: 3776401
object-detector-python-object-detector-python-1  | Dropping old frames: 3276704
object-detector-python-object-detector-python-1  | <_InactiveRpcError of RPC that terminated with:
object-detector-python-object-detector-python-1  |      status = StatusCode.DEADLINE_EXCEEDED
object-detector-python-object-detector-python-1  |      details = "Deadline Exceeded"
object-detector-python-object-detector-python-1  |      debug_error_string = "{"created":"@1659657759.759017906","description":"Error received from peer ipv4:172.20.0.4:8501","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"Deadline Exceeded","grpc_status":4}"
object-detector-python-object-detector-python-1  | >
object-detector-python-object-detector-python-1  | Dropping old frames: 5648180
object-detector-python-object-detector-python-1  | Dropping old frames: 5042697
object-detector-python-object-detector-python-1  | Dropping old frames: 4653912
object-detector-python-object-detector-python-1  | <_InactiveRpcError of RPC that terminated with:
object-detector-python-object-detector-python-1  |      status = StatusCode.DEADLINE_EXCEEDED
object-detector-python-object-detector-python-1  |      details = "Deadline Exceeded"
object-detector-python-object-detector-python-1  |      debug_error_string = "{"created":"@1659657766.692593764","description":"Error received from peer ipv4:172.20.0.4:8501","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"Deadline Exceeded","grpc_status":4}"
object-detector-python-object-detector-python-1  | >
read tcp 10.223.140.46:39406->192.168.10.3:2376: read: connection reset by peer

**Reply by igor-rud-kr** [reply]
I also tried floating point model, it resulted in slightly different error:

object-detector-python-acap_dl-models-1          | detect4.tflite
object-detector-python-acap_dl-models-1          | labels4.txt
object-detector-python-acap_dl-models-1 exited with code 0
object-detector-python-inference-server-1        | Server listening on 0.0.0.0:8501
object-detector-python-object-detector-python-1  | object-detector-python connect to: inference-server:8501
object-detector-python-object-detector-python-1  | <_InactiveRpcError of RPC that terminated with:
object-detector-python-object-detector-python-1  |      status = StatusCode.DEADLINE_EXCEEDED
object-detector-python-object-detector-python-1  |      details = "Deadline Exceeded"
object-detector-python-object-detector-python-1  |      debug_error_string = "{"created":"@1659661886.936877585","description":"Error received from peer ipv4:172.18.0.3:8501","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"Deadline Exceeded","grpc_status":4}"
object-detector-python-object-detector-python-1  | >
object-detector-python-inference-server-1        | ERROR in Inference: Inference request failed (Could not run job: Could only read 270000 out of 1080000 bytes from file descriptor)
object-detector-python-object-detector-python-1  | Dropping old frames: 3618286
object-detector-python-object-detector-python-1  | Dropping old frames: 3120014
object-detector-python-object-detector-python-1  | Dropping old frames: 2617792
object-detector-python-inference-server-1        | ERROR in Inference: Inference request failed (Could not run job: Could only read 270000 out of 1080000 bytes from file descriptor)
object-detector-python-object-detector-python-1  | <_InactiveRpcError of RPC that terminated with:
object-detector-python-object-detector-python-1  |      status = StatusCode.CANCELLED
object-detector-python-object-detector-python-1  |      details = ""
object-detector-python-object-detector-python-1  |      debug_error_string = "{"created":"@1659661888.030394641","description":"Error received from peer ipv4:172.18.0.3:8501","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"","grpc_status":1}"
object-detector-python-object-detector-python-1  | >
object-detector-python-object-detector-python-1  | Dropping old frames: 564164
object-detector-python-inference-server-1        | ERROR in Inference: Inference request failed (Could not run job: Could only read 270000 out of 1080000 bytes from file descriptor)
object-detector-python-object-detector-python-1  | <_InactiveRpcError of RPC that terminated with:
object-detector-python-object-detector-python-1  |      status = StatusCode.CANCELLED
object-detector-python-object-detector-python-1  |      details = ""
object-detector-python-object-detector-python-1  |      debug_error_string = "{"created":"@1659661889.090731232","description":"Error received from peer ipv4:172.18.0.3:8501","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"","grpc_status":1}"
object-detector-python-object-detector-python-1  | >
object-detector-python-object-detector-python-1  | Dropping old frames: 625748
object-detector-python-object-detector-python-1  | Dropping old frames: 125725
object-detector-python-inference-server-1        | ERROR in Inference: Inference request failed (Could not run job: Could only read 270000 out of 1080000 bytes from file descriptor)
object-detector-python-object-detector-python-1  | <_InactiveRpcError of RPC that terminated with:
object-detector-python-object-detector-python-1  |      status = StatusCode.CANCELLED
object-detector-python-object-detector-python-1  |      details = ""
object-detector-python-object-detector-python-1  |      debug_error_string = "{"created":"@1659661890.529351659","description":"Error received from peer ipv4:172.18.0.3:8501","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"","grpc_status":1}"
object-detector-python-object-detector-python-1  | >

**Reply by Corallo** [reply]
Keep using the int quantized version.
After you start the application, let it run, in another shell check with 
`docker -H IP_camera ps` if the inference server container is still running.
And also ssh into the camera and run `journalctl -u larod` what's in the log.
Hopefully this way you can understand better what happens.

**Reply by igor-rud-kr** [reply]
Hi Angelo,

Thank you, I re-tested is with int quantized version. This time it displayed the error messages above for about a minute, but then inference server started working.

**Reply by Corallo** [reply]
Yeah, it happens on artpec8 cameras because before the first inference an internal conversion is executed, it can take up to 1 min.
